File: docs/absorbed-from-phenotype-gfx/unity/postfx/Runtime/PostStack.cs
using UnityEngine;
using Phenotype.PostFx.Ports;

namespace Phenotype.PostFx
{
    // ---------------------------------------------------------------------------
    // Shader-availability abstraction — injectable for tests
    // ---------------------------------------------------------------------------

    /// <summary>
    /// Abstracts shader/material availability so tests can inject mock results
    /// without requiring a live Unity runtime.
    /// </summary>
    public interface IShaderAvailabilityProvider
    {
        /// <summary>
        /// Returns <see langword="true"/> when the shader for the given effect is loaded
        /// and all required variants are present in the build.
        /// </summary>
        /// <param name="effect">The post-processing effect to check.</param>
        /// <returns>
        /// <see langword="true"/> if the shader is available; otherwise, <see langword="false"/>.
        /// </returns>
        bool IsAvailable(PostFxEffect effect);
    }

    /// <summary>Identifies each post-processing effect.</summary>
    public enum PostFxEffect
    {
        /// <summary>Screen Space Ambient Occlusion.</summary>
        SSAO,
        /// <summary>Screen Space Global Illumination.</summary>
        SSGI,
        /// <summary>Bloom glow effect.</summary>
        Bloom,
        /// <summary>ACES tone mapping.</summary>
        ACES,
        /// <summary>Vignette darkening.</summary>
        Vignette,
        /// <summary>Chromatic aberration distortion.</summary>
        ChromaticAberration,
        /// <summary>Color Look-Up Table grading.</summary>
        LUT,
    }

    /// <summary>
    /// Shared per-pass metadata used by the render pipeline.
    /// </summary>
    internal readonly struct PostFxPass
    {
        /// <summary>
        /// Gets the name of the shader used by this pass.
        /// </summary>
        /// <value>The shader name.</value>
        public readonly string ShaderName;

        /// <summary>
        /// Gets the material instance for this pass.
        /// </summary>
        /// <value>The material, or <see langword="null"/> if unavailable.</value>
        public readonly Material? Material;

        /// <summary>
        /// Gets a value indicating whether the pass is enabled.
        /// </summary>
        /// <value>
        /// <see langword="true"/> if enabled; otherwise, <see langword="false"/>.
        /// </value>
        public readonly bool Enabled;

        /// <summary>
        /// Gets a value indicating whether the pass is supported in this build.
        /// </summary>
        /// <value>
        /// <see langword="true"/> if supported; otherwise, <see langword="false"/>.
        /// </value>
        public readonly bool Supported;

        /// <summary>
        /// Initializes a new instance of the <see cref="PostFxPass"/> struct.
        /// </summary>
        /// <param name="shaderName">The shader name.</param>
        /// <param name="material">The material instance.</param>
        /// <param name="enabled">Whether the pass is enabled.</param>
        /// <param name="supported">Whether the pass is supported.</param>
        public PostFxPass(string shaderName, Material? material, bool enabled, bool supported)
        {
            ShaderName = shaderName;
            Material = material;
            Enabled = enabled;
            Supported = supported;
        }

        /// <summary>
        /// Gets a value indicating whether this pass is active.
        /// A pass is active when it is enabled, supported, and has a valid material.
        /// </summary>
        /// <value>
        /// <see langword="true"/> if the pass should execute; otherwise, <see langword="false"/>.
        /// </value>
        public bool IsActive => Enabled && Supported && Material;
    }

    /// <summary>
    /// Production implementation: an effect is available when its material was
    /// successfully loaded (non-null), which is exactly what TryLoad / the LUT
    /// path guarantee at init time.
    /// </summary>
    internal sealed class DefaultShaderAvailabilityProvider : IShaderAvailabilityProvider
    {
        readonly PostStack _owner;

        /// <summary>
        /// Initializes a new instance of the <see cref="DefaultShaderAvailabilityProvider"/> class.
        /// </summary>
        /// <param name="owner">The <see cref="PostStack"/> to query for material availability.</param>
        internal DefaultShaderAvailabilityProvider(PostStack owner) => _owner = owner;

        /// <summary>
        /// Checks whether the specified effect's material is loaded.
        /// </summary>
        /// <param name="effect">The effect to check.</param>
        /// <returns>
        /// <see langword="true"/> if the material is non-null; otherwise, <see langword="false"/>.
        /// </returns>
        public bool IsAvailable(PostFxEffect effect) => effect switch
        {
            PostFxEffect.SSAO  => _owner._ssaoMat  != null,
            PostFxEffect.SSGI  => _owner._ssgiMat  != null,
            PostFxEffect.Bloom => _owner._bloomMat != null,
            PostFxEffect.ACES  => _owner._acesMat  != null,
            PostFxEffect.Vignette => _owner._vignetteMat != null,
            PostFxEffect.ChromaticAberration => _owner._chromaticAberrationMat != null,
            PostFxEffect.LUT   => _owner._lutMat   != null,
            _ => false,
        };
    }

    // ---------------------------------------------------------------------------

    /// <summary>
    /// Central driver for the post-processing stack.
    /// Attach this <see cref="MonoBehaviour"/> to a camera to apply a configurable
    /// chain of post-processing effects (SSAO, Bloom, ACES, Vignette, etc.).
    /// </summary>
    /// <remarks>
    /// <para>
    /// The stack initialises materials on <c>Awake</c>, validates shader variants,
    /// and composites enabled passes via <c>OnRenderImage</c>.  Pass order is
    /// controlled by <see cref="PostFxPassRegistry"/>.
    /// </para>
    /// <para>
    /// Example usage in a scene:
    /// <code>
    /// var stack = camera.gameObject.AddComponent&lt;PostStack&gt;();
    /// stack.EnableBloom = true;
    /// stack.EnableACES = true;
    /// </code>
    /// </para>
    /// </remarks>
    public sealed class PostStack : MonoBehaviour
    {
        /// <summary>
        /// Shader property ID for the bloom texture (<c>_BloomTex</c>).
        /// Used by <see cref="BloomPassProvider"/> to bind the blurred intermediate.
        /// </summary>
        public static readonly int BloomTexId = Shader.PropertyToID("_BloomTex");
        static readonly int ExposureId = Shader.PropertyToID("_Exposure");

        [Header("Pass Toggles")]
        /// <summary>
        /// Enables or disables the Screen Space Ambient Occlusion (SSAO) pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable SSAO; otherwise, <see langword="false"/>.
        /// Default is <see langword="true"/>.
        /// </value>
        public bool EnableSSAO = true;

        /// <summary>
        /// Enables or disables the Screen Space Global Illumination (SSGI) pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable SSGI; otherwise, <see langword="false"/>.
        /// Default is <see langword="false"/>.
        /// </value>
        public bool EnableSSGI;

        /// <summary>
        /// Enables or disables the Bloom pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable Bloom; otherwise, <see langword="false"/>.
        /// Default is <see langword="false"/>.
        /// </value>
        public bool EnableBloom;

        /// <summary>
        /// Enables or disables the ACES tone mapping pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable ACES; otherwise, <see langword="false"/>.
        /// Default is <see langword="true"/>.
        /// </value>
        public bool EnableACES = true;

        /// <summary>
        /// Enables or disables the Vignette pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable Vignette; otherwise, <see langword="false"/>.
        /// Default is <see langword="false"/>.
        /// </value>
        public bool EnableVignette;

        /// <summary>
        /// Enables or disables the Chromatic Aberration pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable Chromatic Aberration; otherwise, <see langword="false"/>.
        /// Default is <see langword="false"/>.
        /// </value>
        public bool EnableChromaticAberration;

        /// <summary>
        /// Enables or disables the Color Look-Up Table (LUT) grading pass.
        /// </summary>
        /// <value>
        /// <see langword="true"/> to enable LUT; otherwise, <see langword="false"/>.
        /// Default is <see langword="true"/>.
        /// </value>
        public bool EnableLUT = true;

        [Header("Quality")]
        /// <summary>
        /// The overall quality preset for the post-processing stack.
        /// </summary>
        /// <value>
        /// A <see cref="PostFxQuality"/> value. Default is <see cref="PostFxQuality.High"/>.
        /// </value>
        public PostFxQuality Quality = PostFxQuality.High;

        [Header("SSAO")]
        /// <summary>
        /// Number of sample points used by the SSAO kernel.
        /// Higher values improve quality at the cost of performance.
        /// </summary>
        /// <value>The sample count. Default is 12.</value>
        public int SSAOSamples = 12;

        /// <summary>
        /// World-space radius of the SSAO sampling sphere.
        /// </summary>
        /// <value>The radius in world units. Default is 2.0f.</value>
        public float SSAORadius = 2.0f;

        /// <summary>
        /// Depth bias for SSAO to avoid self-occlusion artifacts.
        /// </summary>
        /// <value>The bias value. Default is 0.0012f.</value>
        public float SSAOBias = 0.0012f;

        /// <summary>
        /// Intensity multiplier for the SSAO occlusion mask.
        /// </summary>
        /// <value>The intensity. Default is 1.0f.</value>
        public float SSAOIntensity = 1.0f;

        [Header("SSGI")]
        /// <summary>
        /// Number of sample points used by the SSGI kernel.
        /// </summary>
        /// <value>The sample count. Default is 12.</value>
        public int SSGISamples = 12;

        /// <summary>
        /// World-space radius of the SSGI sampling sphere.
        /// </summary>
        /// <value>The radius in world units. Default is 1.8f.</value>
        public float SSGIRadius = 1.8f;

        /// <summary>
        /// Intensity multiplier for the SSGI indirect lighting contribution.
        /// </summary>
        /// <value>The intensity. Default is 0.45f.</value>
        public float SSGIIntensity = 0.45f;

        [Header("ACES")]
        /// <summary>
        /// Exposure value applied before ACES tone mapping.
        /// </summary>
        /// <value>The exposure. Default is 1.0f (no change).</value>
        public float Exposure = 1.0f;

        [Header("Vignette")]
        /// <summary>
        /// Normalized screen-space center of the vignette effect.
        /// </summary>
        /// <value>The center position. Default is (0.5, 0.5).</value>
        public Vector2 VignetteCenter = new Vector2 { x = 0.5f, y = 0.5f };

        /// <summary>
        /// Strength of the vignette darkening.
        /// </summary>
        /// <value>The intensity. Default is 0.45f.</value>
        public float VignetteIntensity = 0.45f;

        /// <summary>
        /// Falloff smoothness of the vignette edge.
        /// </summary>
        /// <value>The smoothness. Default is 0.6f.</value>
        public float VignetteSmoothness = 0.6f;

        /// <summary>
        /// Roundness factor of the vignette shape (0 = square, 1 = circular).
        /// </summary>
        /// <value>The roundness. Default is 1.0f.</value>
        public float VignetteRoundness = 1.0f;

        [Header("Chromatic Aberration")]
        /// <summary>
        /// Strength of the RGB channel separation.
        /// </summary>
        /// <value>The intensity. Default is 0.15f.</value>
        public float ChromaticAberrationIntensity = 0.15f;

        [Header("LUT")]
        /// <summary>
        /// The color grading Look-Up Table texture.
        /// Set to <see langword="null"/> to disable LUT grading.
        /// </summary>
        /// <value>The LUT <see cref="Texture2D"/>.</value>
        public Texture2D LutTexture;

        // Materials — internal so DefaultShaderAvailabilityProvider can read them
        internal Material _ssaoMat;
        internal Material _ssgiMat;
        internal Material _bloomMat;
        internal Material _acesMat;
        internal Material _vignetteMat;
        internal Material _chromaticAberrationMat;
        internal Material _lutMat;

        RenderTexture _ping;
        bool _initialized;
        static readonly Vector4[] _ssaoKernel = new Vector4[16];
        static readonly Vector4[] _ssgiKernel = new Vector4[12];
        static bool _kernelsBuilt;

        // Validated availability flags — set by ValidateShaderVariants()
        internal bool _ssaoSupported;
        internal bool _ssgiSupported;
        internal bool _bloomSupported;
        internal bool _acesSupported;
        internal bool _vignetteSupported;
        internal bool _chromaticAberrationSupported;
        internal bool _lutSupported;

        /// <summary>
        /// Injectable shader-availability provider. Defaults to the production
        /// implementation; replace in tests via <see cref="SetAvailabilityProvider"/>.
        /// </summary>
        IShaderAvailabilityProvider _availabilityProvider;

        /// <summary>
        /// Composio-style pass registry. Replace or extend at runtime to add
        /// custom post-processing passes without modifying <see cref="PostStack"/>.
        /// </summary>
        /// <value>The current pass registry.</value>
        public PostFxPassRegistry PassRegistry { get; private set; } = PostFxPassRegistry.CreateDefault();

        /// <summary>
        /// Initializes a new instance of the <see cref="PostStack"/> class.
        /// Binds the owner to <see cref="PassRegistry"/> so providers that cache
        /// the owner (e.g. <see cref="BlitPassProvider"/>) have it before
        /// <c>Awake</c> or <c>InitMaterials</c> run.
        /// </summary>
        public PostStack()
        {
            PassRegistry.Init(this);
        }

        // ------------------------------------------------------------------
        // Public API for test injection
        // ------------------------------------------------------------------

        /// <summary>
        /// Overrides the shader-availability provider used by
        /// <see cref="ValidateShaderVariants"/>.
        /// </summary>
        /// <param name="provider">The provider to use.</param>
        /// <exception cref="System.ArgumentNullException">
        /// Thrown when <paramref name="provider"/> is <see langword="null"/>.
        /// </exception>
        /// <remarks>
        /// Call before <c>Awake</c> or immediately after construction in tests.
        /// </remarks>
        public void SetAvailabilityProvider(IShaderAvailabilityProvider provider)
            => _availabilityProvider = provider ?? throw new System.ArgumentNullException(nameof(provider));

        // ------------------------------------------------------------------

        void Awake()
        {
            Camera cam = GetComponent<Camera>();
            if (cam != null) cam.depthTextureMode |= DepthTextureMode.Depth;
            BuildKernels();
            InitMaterials();
        }

        void OnDestroy()
        {
            ReleaseMaterials();
            ReleasePingPong();
            PassRegistry.Dispose();
        }

        static void BuildKernels()
        {
            if (_kernelsBuilt) return;
            var rng = new System.Random(1337);
            for (int i = 0; i < _ssaoKernel.Length; i++)
            {
                Vector2 s;
                s.x = Mathf.Lerp(-1f, 1f, (float)rng.NextDouble());
                s.y = Mathf.Lerp(-1f, 1f, (float)rng.NextDouble());
                if (s.sqrMagnitude < 0.0001f) s = Vector2.right;
                s.Normalize();
                _ssaoKernel[i] = new Vector4(s.x, s.y, 0f, (i + 1f) / _ssaoKernel.Length);
            }
            rng = new System.Random(4242);
            for (int i = 0; i < _ssgiKernel.Length; i++)
            {
                Vector2 s;
                s.x = Mathf.Lerp(-1f, 1f, (float)rng.NextDouble());
                s.y = Mathf.Lerp(-1f, 1f, (float)rng.NextDouble());
                if (s.sqrMagnitude < 0.0001f) s = Vector2.one;
                s.Normalize();
                _ssgiKernel[i] = new Vector4(s.x, s.y, 0f, 0f);
            }
            _kernelsBuilt = true;
        }

        void InitMaterials()
        {
            ReleaseMaterials();
            _ssaoMat = TryLoadPass("Shaders/ScreenSpaceAO", "Hidden/ScreenSpaceAO");
            _ssgiMat = TryLoadPass("Shaders/ScreenSpaceGI", "Hidden/ScreenSpaceGI");
            _bloomMat = TryLoadPass("Shaders/BrpBloom", "Hidden/Phenotype/BrpBloom");
            _acesMat = TryLoadPass("Shaders/BrpACES", "Hidden/Phenotype/BrpACES");
            _vignetteMat = TryLoadPass("Shaders/Vignette", "Hidden/WSM3D/Vignette");
            _chromaticAberrationMat = TryLoadPass("Shaders/ChromaticAberration", "Hidden/WSM3D/ChromaticAberration");

            Shader lutShader = Resources.Load<Shader>("Shaders/ColorGradingLUT");
            lutShader ??= Shader.Find("Hidden/ColorGradingLUT");
            if (lutShader != null)
            {
                _lutMat = new Material(lutShader);
                if (LutTexture != null)
                {
                    if (_lutMat.HasProperty("_LutTex")) _lutMat.SetTexture("_LutTex", LutTexture);
                    else if (_lutMat.HasProperty("_LookupTex")) _lutMat.SetTexture("_LookupTex", LutTexture);
                }
                if (_lutMat.HasProperty("_LutParams"))
                    _lutMat.SetVector("_LutParams", new Vector4(16f / 256f, 1f / 16f, 1f, 0f));
            }

            if (_ssaoMat != null) ApplySSAOParams();
            if (_ssgiMat != null) ApplySSGIParams();

            // Resolve the provider lazily so tests can inject before Awake()
            _availabilityProvider ??= new DefaultShaderAvailabilityProvider(this);
            ValidateShaderVariants();

            _initialized = true;
        }

        // ------------------------------------------------------------------
        // Shader-variant validation
        // ------------------------------------------------------------------

        /// <summary>
        /// Audits each enabled effect against the availability provider.
        /// On failure the effect is silently disabled for this session and a
        /// <see cref="Debug.LogWarning"/> is emitted — preventing black/pink
        /// renders in stripped builds.
        /// </summary>
        internal void ValidateShaderVariants()
        {
            _ssaoSupported  = ValidatePass(PostFxEffect.SSAO,  "ScreenSpaceAO",  nameof(EnableSSAO));
            _ssgiSupported  = ValidatePass(PostFxEffect.SSGI,  "ScreenSpaceGI",  nameof(EnableSSGI));
            _bloomSupported = ValidatePass(PostFxEffect.Bloom, "BrpBloom",       nameof(EnableBloom));
            _acesSupported  = ValidatePass(PostFxEffect.ACES,  "BrpACES",        nameof(EnableACES));
            _vignetteSupported = ValidatePass(PostFxEffect.Vignette, "Vignette", nameof(EnableVignette));
            _chromaticAberrationSupported = ValidatePass(PostFxEffect.ChromaticAberration, "ChromaticAberration", nameof(EnableChromaticAberration));
            _lutSupported   = ValidatePass(PostFxEffect.LUT,   "ColorGradingLUT",nameof(EnableLUT));
        }

        bool ValidatePass(PostFxEffect effect, string shaderName, string toggleName)
        {
            bool available = _availabilityProvider.IsAvailable(effect);
            if (!available)
            {
                Debug.LogWarning(
                    $"[PostStack] Shader variant unavailable: {shaderName}. " +
                    $"{toggleName} will be skipped this session. " +
                    $"Ensure the shader is included in your build's ShaderVariantCollection " +
                    $"(phenotype-postfx-variants.shadervariants).");
            }
            return available;
        }

        void ReleaseMaterials()
        {
            if (_ssaoMat != null) { Destroy(_ssaoMat); _ssaoMat = null; }
            if (_ssgiMat != null) { Destroy(_ssgiMat); _ssgiMat = null; }
            if (_bloomMat != null) { Destroy(_bloomMat); _bloomMat = null; }
            if (_acesMat != null) { Destroy(_acesMat); _acesMat = null; }
            if (_vignetteMat != null) { Destroy(_vignetteMat); _vignetteMat = null; }
            if (_chromaticAberrationMat != null) { Destroy(_chromaticAberrationMat); _chromaticAberrationMat = null; }
            if (_lutMat != null) { Destroy(_lutMat); _lutMat = null; }
        }

        static Material? TryLoadPass(string resourcePath, string fallbackName)
        {
            Shader shader = Resources.Load<Shader>(resourcePath);
            shader ??= Shader.Find(fallbackName);
            return shader != null ? new Material(shader) : null;
        }

        internal void ApplySSAOParams()
        {
            _ssaoMat.SetInt("_SampleCount", SSAOSamples);
            _ssaoMat.SetVectorArray("_Samples", _ssaoKernel);
            _ssaoMat.SetFloat("_Radius", SSAORadius);
            _ssaoMat.SetFloat("_Bias", SSAOBias);
            _ssaoMat.SetFloat("_Intensity", SSAOIntensity);
        }

        internal void ApplySSGIParams()
        {
            _ssgiMat.SetInt("_SampleCount", SSGISamples);
            _ssgiMat.SetVectorArray("_Samples", _ssgiKernel);
            _ssgiMat.SetFloat("_Radius", SSGIRadius);
            _ssgiMat.SetFloat("_Intensity", SSGIIntensity);
        }

        internal void ApplyACESParams()
        {
            _acesMat.SetFloat(ExposureId, Exposure);
        }

        internal void ApplyVignetteParams()
        {
            _vignetteMat.SetVector("_Center", new Vector4(VignetteCenter.x, VignetteCenter.y, 0f, 0f));
            _vignetteMat.SetFloat("_Intensity", VignetteIntensity);
            _vignetteMat.SetFloat("_Smoothness", VignetteSmoothness);
            _vignetteMat.SetFloat("_Roundness", VignetteRoundness);
        }

        internal void ApplyChromaticAberrationParams()
        {
            _chromaticAberrationMat.SetFloat("_Intensity", ChromaticAberrationIntensity);
        }

        void EnsurePingPong(RenderTexture src)
        {
            if (_ping != null && _ping.width == src.width && _ping.height == src.height) return;
            ReleasePingPong();
            _ping = RenderTexture.GetTemporary(src.descriptor);
        }

        void ReleasePingPong()
        {
            if (_ping != null) { RenderTexture.ReleaseTemporary(_ping); _ping = null; }
        }

        static bool BlitIfEnabled(RenderTexture src, RenderTexture dst, in PostFxPass pass, System.Action? beforeBlit = null)
        {
            if (!pass.IsActive) return false;
            beforeBlit?.Invoke();
            Graphics.Blit(src, dst, pass.Material!);
            return true;
        }

        void OnRenderImage(RenderTexture src, RenderTexture dst)
        {
            if (src == null || dst == null) return;
            if (!_initialized)
            {
                Graphics.Blit(src, dst);
                return;
            }

            if (!PassRegistry.HasAnyActivePass(this))
            {
                Graphics.Blit(src, dst);
                return;
            }

            EnsurePingPong(src);
            try
            {
                RenderTexture cur = src, next = _ping;

                foreach (var effect in PassRegistry.Providers)
                {
                    if (!effect.IsEnabled(this) || !effect.IsSupported(this))
                        continue;

                    // LUT is applied as a final composite pass to dst
                    if (effect.Effect == PostFxEffect.LUT)
                    {
                        if (LutTexture != null && effect.Material != null)
                            Graphics.Blit(cur, dst, effect.Material);
                        else
                            Graphics.Blit(cur, dst);
                        continue;
                    }

                    effect.ApplyParams(this);
                    if (effect.Render(cur, next))
                    {
                        Swap(ref cur, ref next);
                    }
                }

                // If LUT was not active, blit the final result to dst
                var lutProvider = PassRegistry.GetProvider(PostFxEffect.LUT);
                if (lutProvider == null || !lutProvider.IsEnabled(this) || !lutProvider.IsSupported(this) || LutTexture == null)
                {
                    Graphics.Blit(cur, dst);
                }
            }
            finally
            {
                ReleasePingPong();
            }
        }

        static void Swap(ref RenderTexture a, ref RenderTexture b) => (a, b) = (b, a);
    }
}

File: docs/absorbed-from-phenotype-gfx/src/postfx/post_stack.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
// SPDX-FileCopyrightText: 2026 KooshaPari <kooshapari@gmail.com>

//! `PostStack` — central driver for the post-processing stack.
//!
//! Engine-agnostic port from the C# `PostStack.cs` (MonoBehaviour). The Rust
//! side holds the configuration, the validated availability flags, and the
//! pass registry. The C# edge binds the registry to a `MonoBehaviour` at
//! runtime and dispatches `on_render` for each pass in render order.
//!
//! The `describe_passes()` method returns a `Vec<PassDescriptor>` so that
//! the editor / driver can build an inspector / dispatcher without referring
//! to the concrete `BloomPass` / `SsaoPass` types.

use serde::{Deserialize, Serialize};

use crate::postfx::bloom_pass::BloomConfig;
use crate::postfx::ports::post_fx_pass::{
    PassDescriptor, PassEffect, PassQuality, PostFxContext, PostFxPass,
};
use crate::postfx::ports::shader_availability::PostFxShaderAvailability;
use crate::postfx::post_fx_pass_registry::PostFxPassRegistry;
use crate::postfx::ssao_pass::SsaoConfig;

/// All post-fx configuration. Engine-agnostic: carries only logical state,
/// not Unity references.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PostStackConfig {
    /// SSAO toggle.
    pub enable_ssao: bool,
    /// SSGI toggle.
    pub enable_ssgi: bool,
    /// Bloom toggle.
    pub enable_bloom: bool,
    /// ACES toggle.
    pub enable_aces: bool,
    /// Vignette toggle.
    pub enable_vignette: bool,
    /// Chromatic aberration toggle.
    pub enable_chromatic_aberration: bool,
    /// LUT toggle.
    pub enable_lut: bool,

    /// Overall quality preset.
    pub quality: PassQuality,

    /// SSAO sample count.
    pub ssao_samples: u32,
    /// SSAO world-space radius.
    pub ssao_radius: f32,
    /// SSAO depth bias.
    pub ssao_bias: f32,
    /// SSAO intensity.
    pub ssao_intensity: f32,

    /// SSGI sample count.
    pub ssgi_samples: u32,
    /// SSGI world-space radius.
    pub ssgi_radius: f32,
    /// SSGI intensity.
    pub ssgi_intensity: f32,

    /// Exposure for ACES.
    pub exposure: f32,
    /// Vignette intensity.
    pub vignette_intensity: f32,
    /// Vignette smoothness.
    pub vignette_smoothness: f32,
    /// Vignette roundness.
    pub vignette_roundness: f32,
    /// Vignette center (normalized).
    pub vignette_center: [f32; 2],
    /// Chromatic aberration intensity.
    pub chromatic_aberration_intensity: f32,
}

impl Default for PostStackConfig {
    fn default() -> Self {
        Self {
            enable_ssao: true,
            enable_ssgi: false,
            enable_bloom: false,
            enable_aces: true,
            enable_vignette: false,
            enable_chromatic_aberration: false,
            enable_lut: true,
            quality: PassQuality::High,
            ssao_samples: 12,
            ssao_radius: 2.0,
            ssao_bias: 0.0012,
            ssao_intensity: 1.0,
            ssgi_samples: 12,
            ssgi_radius: 1.8,
            ssgi_intensity: 0.45,
            exposure: 1.0,
            vignette_intensity: 0.45,
            vignette_smoothness: 0.6,
            vignette_roundness: 1.0,
            vignette_center: [0.5, 0.5],
            chromatic_aberration_intensity: 0.15,
        }
    }
}

impl PostStackConfig {
    /// Build the bloom sub-config from the post-stack config.
    pub fn bloom_config(&self) -> BloomConfig {
        BloomConfig {
            is_enabled: self.enable_bloom,
            ..BloomConfig::default()
        }
    }

    /// Build the SSAO sub-config from the post-stack config.
    pub fn ssao_config(&self) -> SsaoConfig {
        SsaoConfig {
            is_enabled: self.enable_ssao,
            radius: self.ssao_radius,
            intensity: self.ssao_intensity,
            bias: self.ssao_bias,
            kernel_size: self.ssao_samples,
        }
    }
}

/// Central driver for the post-processing stack.
///
/// Engine-agnostic; the C# edge wraps this in a `MonoBehaviour` and calls
/// `on_render` for each pass during `OnRenderImage`.
pub struct PostStack {
    config: PostStackConfig,
    registry: PostFxPassRegistry,
    /// Validated availability flags — set by `validate_shader_variants`.
    ssao_supported: bool,
    ssgi_supported: bool,
    bloom_supported: bool,
    aces_supported: bool,
    vignette_supported: bool,
    chromatic_aberration_supported: bool,
    lut_supported: bool,
}

/// Default-constructed post-stack used as the canonical config baseline.
pub const DEFAULT_POSTFX_STACK: PostStackConfig = PostStackConfig {
    enable_ssao: true,
    enable_ssgi: false,
    enable_bloom: false,
    enable_aces: true,
    enable_vignette: false,
    enable_chromatic_aberration: false,
    enable_lut: true,
    quality: PassQuality::High,
    ssao_samples: 12,
    ssao_radius: 2.0,
    ssao_bias: 0.0012,
    ssao_intensity: 1.0,
    ssgi_samples: 12,
    ssgi_radius: 1.8,
    ssgi_intensity: 0.45,
    exposure: 1.0,
    vignette_intensity: 0.45,
    vignette_smoothness: 0.6,
    vignette_roundness: 1.0,
    vignette_center: [0.5, 0.5],
    chromatic_aberration_intensity: 0.15,
};

impl PostStack {
    /// New post-stack with the given config and an empty registry.
    pub fn new(config: PostStackConfig) -> Self {
        Self {
            config,
            registry: PostFxPassRegistry::new(),
            ssao_supported: false,
            ssgi_supported: false,
            bloom_supported: false,
            aces_supported: false,
            vignette_supported: false,
            chromatic_aberration_supported: false,
            lut_supported: false,
        }
    }

    /// Borrow the current config.
    pub fn config(&self) -> &PostStackConfig {
        &self.config
    }

    /// Mutably borrow the current config.
    pub fn config_mut(&mut self) -> &mut PostStackConfig {
        &mut self.config
    }

    /// Borrow the registry.
    pub fn registry(&self) -> &PostFxPassRegistry {
        &self.registry
    }

    /// Mutably borrow the registry.
    pub fn registry_mut(&mut self) -> &mut PostFxPassRegistry {
        &mut self.registry
    }

    /// Returns the static descriptors for the built-in passes, in the
    /// canonical render order. The driver / editor uses this to build an
    /// inspector / dispatcher without referring to the concrete `BloomPass`
    /// / `SsaoPass` types.
    pub fn describe_passes() -> Vec<PassDescriptor> {
        vec![
            SsaoConfig::descriptor(),
            BloomConfig::descriptor(),
        ]
    }

    /// Audits each effect against the availability provider and updates the
    /// per-effect support flags.
    pub fn validate_shader_variants(&mut self, provider: &dyn PostFxShaderAvailability) {
        self.ssao_supported =
            provider.is_available("ScreenSpaceAO", "SSAOPASS");
        self.ssgi_supported =
            provider.is_available("ScreenSpaceGI", "SSGIPASS");
        self.bloom_supported =
            provider.is_available("BrpBloom", "BLOOM_HIGH");
        self.aces_supported =
            provider.is_available("BrpACES", "ACES");
        self.vignette_supported =
            provider.is_available("Vignette", "VIGNETTE");
        self.chromatic_aberration_supported =
            provider.is_available("ChromaticAberration", "CHROMATIC");
        self.lut_supported =
            provider.is_available("ColorGradingLUT", "LUT");
    }

    /// Returns `true` if the SSAO shader variant is available.
    pub fn ssao_supported(&self) -> bool { self.ssao_supported }
    /// Returns `true` if the SSGI shader variant is available.
    pub fn ssgi_supported(&self) -> bool { self.ssgi_supported }
    /// Returns `true` if the Bloom shader variant is available.
    pub fn bloom_supported(&self) -> bool { self.bloom_supported }
    /// Returns `true` if the ACES shader variant is available.
    pub fn aces_supported(&self) -> bool { self.aces_supported }
    /// Returns `true` if the Vignette shader variant is available.
    pub fn vignette_supported(&self) -> bool { self.vignette_supported }
    /// Returns `true` if the ChromaticAberration shader variant is available.
    pub fn chromatic_aberration_supported(&self) -> bool { self.chromatic_aberration_supported }
    /// Returns `true` if the LUT shader variant is available.
    pub fn lut_supported(&self) -> bool { self.lut_supported }
}

impl Default for PostStack {
    fn default() -> Self {
        Self::new(PostStackConfig::default())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::postfx::ports::shader_availability::{DefaultPostFxShaderAvailability, MapPostFxShaderAvailability};

    #[test]
    fn default_config_baseline() {
        let s = PostStack::default();
        let cfg = s.config();
        assert!(cfg.enable_ssao);
        assert!(!cfg.enable_ssgi);
        assert!(!cfg.enable_bloom);
        assert!(cfg.enable_aces);
        assert!(!cfg.enable_vignette);
        assert!(!cfg.enable_chromatic_aberration);
        assert!(cfg.enable_lut);
        assert_eq!(cfg.quality, PassQuality::High);
    }

    #[test]
    fn describe_passes_includes_built_ins() {
        let descs = PostStack::describe_passes();
        assert!(descs.iter().any(|d| d.effect == PassEffect::Bloom));
        assert!(descs.iter().any(|d| d.effect == PassEffect::Ssao));
    }

    #[test]
    fn validate_all_supported_with_default() {
        let mut s = PostStack::default();
        let p = DefaultPostFxShaderAvailability;
        s.validate_shader_variants(&p);
        assert!(s.ssao_supported());
        assert!(s.bloom_supported());
        assert!(s.aces_supported());
        assert!(s.vignette_supported());
        assert!(s.chromatic_aberration_supported());
        assert!(s.lut_supported());
        assert!(s.ssgi_supported());
    }

    #[test]
    fn validate_with_none_available() {
        let mut s = PostStack::default();
        let p = MapPostFxShaderAvailability::none_available();
        s.validate_shader_variants(&*p);
        assert!(!s.ssao_supported());
        assert!(!s.ssgi_supported());
        assert!(!s.bloom_supported());
        assert!(!s.aces_supported());
        assert!(!s.vignette_supported());
        assert!(!s.chromatic_aberration_supported());
        assert!(!s.lut_supported());
    }

    #[test]
    fn bloom_and_ssao_config_derived() {
        let cfg = PostStackConfig::default();
        let b = cfg.bloom_config();
        assert!(!b.is_enabled);
        let s = cfg.ssao_config();
        assert!(s.is_enabled);
        assert_eq!(s.kernel_size, 12);
    }

    #[test]
    fn default_postfx_stack_matches_default_config() {
        assert_eq!(PostStackConfig::default().enable_ssao, DEFAULT_POSTFX_STACK.enable_ssao);
        assert_eq!(PostStackConfig::default().quality, DEFAULT_POSTFX_STACK.quality);
    }
}
